### ISSUE TITLE
Add Donation URLs to Mod Metadata

### ISSUE THREAD

**Issue by NOTtheMessiah** [KSP-CKAN/CKAN#838]
[Maxmaps:](https://www.reddit.com/r/IAmA/comments/33ys4s/we_are_the_team_behind_kerbal_space_program/cqpofr1)
> Our current policy doesn't allow charging for mods, but we encourage people to donate to their favorite mod authors.

I suspect the biggest hurdle towards donating to modders is the inconvenience of the process, and by making a donation link more visible it takes one step out of the process of funding a developer. Curseforge already has donation buttons, and by including it in the metadata, you can use it with CKAN plugins.

Potential additional ideas:
* Multiple donation links ordered by preference.
* Plugin that gathers Bitcoin donation links from installed mods to make a multi-output transaction so you can donate to all your mods at the same time.

**Comment by TeddyDD** [comment]
:+1: for this. Nothing to add :)

**Comment by pjf** [comment]
Absolutely solid request. Popping on our To-do list.

**Comment by pjf** [comment]
Implementation note, this would go in the `resources` section of our metadata, which is absolutely intended to be expanded in this fashion. (Although gathering bitcoin data would be hard, and likely a separate issue if deemed appropriate.)

**Comment by WazWaz** [comment]
Users would need to be able to trust that the donation links are for the person(s) who made the mod, not just someone packaging it up. How can this be "policed"? I can understand why many mod authors don't like CKAN given this - many users only ever go to the forum links when they have a problem - so buggier mods get more hits to the forums.

**Comment by politas** [comment]
I don't think that last has anything to do with CKAN, really. It's an unfortunate reality that people complain and criticise far more than they applaud.

But your point about the verifiability is a good one. How much effort does Curseforge make in policing mod authorship?

**Comment by WazWaz** [comment]
I guess I just can't think of any other reason mod authors wouldn't like CKAN - it seems like the only way to force people to install prerequisites and to keep up-to-date before complaining about bugs. And yet quite a few mod authors do complain - and let's face it: without mod authors, CKAN would be useless. So working out why they don't like it and resolving the problems seems pretty important.

**Comment by politas** [comment]
I think the main problem mod authors have is that sometimes, CKAN gets the prerequisites and installation wrong, because our metadata is incorrect. That's actually a valid complaint, and one we are trying to address.

**Comment by solarsootysmudge** [comment]
@WazWaz Remember that the people that complain about CKAN. Don't have an actually working relationship with the project. Mods can get indexed without input from their authors. They can find this confusing and jump to the wrong conclusions. Despite CKAN having tutorial documentation on the matter. 

If problems are encountered people can complain all day on mod forums about CKAN being wrong and nothing will improve. Quite often all that is required to fix a complain was a NetKAN Issue Report (a PR is even better). Or even a report of the problem in the CKAN forum. Honestly most problems get fixed very quickly once they are communicated directly to CKAN team. It is just easier for people complain that to join in and work the problems together as a community.

To get back on topic. As for including donation links. I think we need to see a stable mod hosting environment with donation links first. Remembering that if CKAN is just scraping data. It needs a good source of that data to be available first. It is still early days for Spacedock and Curse is still presenting a unique problem in redirecting urls. The mod hosting sites need to clarify things first before CKAN gets involved with sharing donation links. Currently CKAN does link to almost every mods forum thread and owners can post donation links there. Although perhaps not as obvious as "button" in CKAN. Doing it this way does provide an element of security and public monitoring through KSP forum controls.

**Comment by plague006** [comment]
The risk vs benefit doesn't add up to me.

The benefit is negligible. If someone has the impetus to donate money to a mod author, having to go to a forum page and then donation page isn't likely to stop them. It also seems like there isn't a huge amount of support for this considering the issue hasn't seen attention in over a year.

By contrast the risk is being implicated in any wrongdoing regarding monetary donations to authors is huge. The likelihood of something going afoul is low, but the repercussions if it did happen would likely be severe. CKAN already receives a lot of criticism for many things we do, adding to our liabilities by involving ourselves in financial dealings seems a risk too far.

**Comment by ferram4** [comment]
On the other hand, there are people who would consider donating if they were aware that the mod author accepted donations, but wouldn't think to seek out a link or even consider donating if the link was not there (out of sight, out of mind).  That means that if a donation link is listed on a forum thread, Curse, or Spacedock, but then isn't listed in CKAN, that represents some small (but non-zero) loss of donations from those users.

**Comment by politas** [comment]
Ok, @ferram4 , but how would you feel, as a mod author who doesn't support CKAN, to find that CKAN was directing donations for one of your mods to someone else who had put up a blatant ripoff on Github? I'm thinking you would be rightfully furious.

**Comment by ferram4** [comment]
And under the current situation, as a mod author that doesn't support CKAN, I find that CKAN is hiding the option to donate from many users that would see it if they went through the supported Spacedock or forum channels.  While someone profiting off another modder's work illicitly is obviously far more insulting and heinous, the current situation has CKAN hiding away the option entirely.  And unless your argument is that there are no people that would donate if they saw a link, but wouldn't think to seek out the option to donate, CKAN must result in some drop in donations through the current setup.  Is that acceptable?

**Comment by janbrohl** [comment]
Ripoffs should not be a big problem as they would not be listed for a long time. When the original author finds out about it or is informed he can file an issue.

It is quite easy to find out who posted the mod first on the forums.

**Comment by mheguy** [comment]
I created a false dichotomy in my assessment. It doesn't have to be "link directly to a paypal/patreon" or "do nothing at all". @ferram4, would it be a fair compromise if we had an indicator of some sort simply indicating that an author has a method of accepting donations while not linking to it directly?

I stand by my reasoning that someone who wants to donate won't be dissuaded by taking 1 extra step, but by simply indicating that an author accepts donations we can effectively increase the number of users who think about donating.

**Comment by janbrohl** [comment]
An option would be to set a flag and link to the forum thread.
edit: That would force mod authors to have a forum thread so it doesn't work :(

**Comment by ferram4** [comment]
@plague006 , I can think of three problems there:
1. This is roundabout and silly; there will be lots of issues created about making the flag an actual link because that would be more straightforward.
2. The concern over the donation link being legit doesn't go away; now it extends to whether the forum thread / Curse page / Spacedock page / etc. has a legit donation link.  People will immediately go to those rather than searching, and now you've got the problem of policing forum threads and hosts that you have no control over.
3. Donation links don't change often.  Forum threads and hosts can, simply because of cleaning up threads or anything.  If the donation link is in the metadata (probably NetKAN is good) then it's obvious if someone tries to change it.  If the donation link is in the forum thread or at the host site and those change, it's less obvious that someone's up to no good with redirecting users to the wrong thread or host.

I mean, if you've got the manpower to dedicate to doing that, by all means, go ahead.  But I'd think keeping track of the known valid donation links and ensuring they aren't touched is far easier, and it removes a barrier to donating.

**Comment by mheguy** [comment]
You're right that if we hardcode donation links it's easy to keep track of from there, the difficulty is in getting that information. Do we have to check every mod for a donation link? Do we force mod authors to come to us? How do we contact all mod authors?

**Comment by WazWaz** [comment]
There is a related problem that is perhaps also part of the solution: as said here, many people do not even realize that some mods available through CKAN are not put there by the authors of the mod; I was surprised, and a little embarassed when I pointed out a CKAN issue to a mod developer who responded "I don't have anything to do with that". It never occurred to me that someone would publish the work of others... until I thought about it and realized CKAN is a crowdsourced directory, not a publishing point, and it all made sense. By I don't think it's obvious to many people. CKAN is so easy to use, I expect many people don't read the tutorial.

So perhaps the solution to the validity of a donation link is flagging "official" CKAN registration - which also hints to users that not all CKAN entries are authored by the mod author.

In the end, I think this would encourage mod developers to use CKAN, which will benefit CKAN users, and I would hope eventually, the mod authors too, because CKAN is a lot easier for the average person to get an install correct (assuming the metadata is correct for the mod they want to install), so there will be more users, and as @ferram4 says, more users means more donations (and kudos and whatever else an author is motivated by).

**Comment by janbrohl** [comment]
I think it would be good to distinguish mods added by the author(s)/maintainer(s) from those added by unrelated people anyway.

edit: created new issue for this

**Comment by ferram4** [comment]
@plague006 , ideal is this: check if the mod has a donate link anywhere when setting it up; if it does, get in touch with the mod author to make sure that the link is correct as part of creating the initial NetKAN; that then gives you the opportunity to tell them that if the link changes the author needs to bring it to you guys.  Simple, easy.  Also can be done starting with what is already available, just with a backlog to get through.

Requiring authors to bring it to your attention in the first place is basically setting up what we have right now until someone finds out and gets annoyed.  Basically, the same problem with it not being there in the first place.

If you can't get in contact with them but they do have a link, then you have to decide whether it's better to list it on CKAN and use that link without verifying, list it on CKAN and use no link (and result in annoyance at hiding donate button) or to go whole-hog on risk-reduction and not list them.

But ultimately, if you don't pick the last option (which we all know is out of the cards) then you fall back to what you have now.  Otherwise, you make the best attempt possible to not interfere with donations.

@WazWaz ...there's kinda a huge problem with that idea.  To users and CKAN contributors, this is providing an incentive to modders.  But to me, where I know I'd get donations if the button is visible but won't if it isn't, it's equivalent to CKAN saying, "we've listed your mod, and now unless you maintain the metadata, we'll make sure you don't get any donations."  I think that might not go over well.

**Comment by WazWaz** [comment]
@ferram4 - I guess I'm still trying to understand why mod authors wouldn't _want_ to maintain the metadata themselves - it only changes when they do something in their release process. This issue seemed one reason, but clearly there are others.

**Comment by mheguy** [comment]
@WazWaz because they don't want to take on extra work? They'd rather be modding?

**Comment by WazWaz** [comment]
@plague006 - that reason would equally apply to releasing the mod at all. FOSS people release stuff to get feedback. CKAN = more feedback just as a forum post does.

**Comment by ferram4** [comment]
@WazWaz More feedback << Good feedback.

CKAN also means more mods involved (more confounders), more deviation from the distributed package (because of updating dependencies faster and very high possibility of incorrect metadata), a userbase that is tends to be less computer literate and aware of what is going on in the background (because it is a video game and CKAN tries to make things easy, and because hiding how the mod is installed is part of what it does), so CKAN really doesn't help me in terms of feedback at all.  I used to get much better feedback before CKAN existed because my userbase was more literate and were actually playing what I released.

**Comment by WazWaz** [comment]
@ferram4 - it's sad, but I can see how this is the case (it is with just about every other thing that becomes popularized). I guess then CKAN is to KSP as Ubuntu/Debian/Canonical is to Linux.

**Comment by ayan4m1** [comment]
After reading through this thread - my two cents - the risk unfortunately outweighs the benefits. If there is still a big push for this, then I think the resources section of the schema could have a "donation" type added to it, but the threat that @plague006 brings up about legal issues is very real. If we disclaimed trust of those links somewhere, then adding a donation link type seems like the best solution. Any kind of automated scraping or parsing of any BitCoin or donation site link seems like a beyond-dangerous and silly idea to me.

**Comment by HebaruSan** [comment]
Two more cents - wouldn't each author use the same donation link for all of his or her mods? This makes me think the donation link doesn't belong under `resources`, but rather in a new author "table" tracked separately. I.e., there should be a single place specifying the donation link for ferram4, and the GUI app should display it for all of the mods that have ferram4 as one of the authors by cross-referencing those lists. This would make for much easier maintenance if an author's link changes, as compared to having to change one file per mod (if it lives in `.netkan`) or one file per version per mod (if it lives in `.ckan`).

Even so, I can think of a scenario in which `resources` would still be the best solution: A team-authored mod with a collective donation link from which the authors split the proceeds. Does such a project exist today?

**Comment by HebaruSan** [comment]
There's always the low tech option of adding a label to the UI:

> REMINDER: Many mods have donation links on their home pages. Be sure to click them to support your favorite modders!

**Comment by HebaruSan** [comment]
The KSP community has seen a series of incidents in which certain mod adopters and maintainers such as @zer0Kerbal low-effort "adopted" mods in the apparent hope of collecting donations for others' work. Adding a `resources.donations` property would be a move toward enabling such abuses rather than curtailing them in order to reward the original creators.

Probably better not to try to ride this tiger.
